source/common/http/conn_manager_utility.cc

```diff
@@ -303,7 +304,19 @@ void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_
 void ConnectionManagerUtility::mutateResponseHeaders(Http::HeaderMap& response_headers,
                                                      const Http::HeaderMap& request_headers,
                                                      const std::string& via) {
-  response_headers.removeConnection();
+  if (Utility::isUpgrade(request_headers) && Utility::isUpgrade(response_headers)) {
+    // As in mutateRequestHeaders, Upgrade responses have special handling.
+    //
+    // Unlike mutateRequestHeaders there is no explicit protocol check. If Envoy is proxying an
+    // upgrade response it has already passed the protocol checks.
+    const bool no_body =
+        (!response_headers.TransferEncoding() && !response_headers.ContentLength());
+    if (no_body) {
+      response_headers.insertContentLength().value(uint64_t(0));
```

Comment:
Why is this needed in the upgrade path but not other cases? I think I'm just ignorant of the standard here...

Reply:
For response headers I believe you can frame by connection close.  That won't work for WebSocket upgrades because you can't close between HTTP and WebSocket payload.